driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -310,14 +307,9 @@ private long waitForSignalOrTimeout() throws InterruptedException {
         }
 
         public void cancelCurrentCheck() {
```

Comment:
Again this looks like it was purposely defensive - any ideas why? 

Could `cancelCurrentCheck` and other usages of `connection` be interweaved - hence requiring the synchronization?

Reply:
> Again this looks like it was purposely defensive - any ideas why?

I can't reconstruct the ideas of why `synchronized` was added because, as far as I can see, it achieves nothing.

> Could cancelCurrentCheck and other usages of connection be interweaved - hence requiring the synchronization?

The only thing that matters for this PR is whether `synchronized` as it is in the code, gives us guarantees that we don't have without it, or not. As far as I can see, it does not. If there is any code that must not be run concurrently with the code inside the `synchronized` block, and that code is not itself guarded by the `synchronized` block using the same `this`, then it is irrelevant whether this `synchronized` block exists or not.

Thus, we only need to consider whether it is fine for the code in the block to run concurrently with itself and with the code in the other `synchronized` block. I convinced myself that it is, and asked reviewers to do the analysis on their own, and bring up specific issues if they see them.

Reply:
Yeah, I agree 👍

Reply:
Currently, `ServerMonitor.cancelCurrentCheck` is not called concurrently with itself, i.e., regardless of whether it's `synchronized` or not, it does not run concurrently with itself. This, in turn, means that `ServerMonitor.cancelCurrentCheck` alone can't result in `ServerMonitorRunnable.connection.close` being called concurrently with itself.

However, if `ServerMonitor.cancelCurrentCheck` is called concurrently with `ServerMonitorRunnable.run`, which is totally possible, then not having the two `synchronized` blocks that were removed, allows concurrently running `ServerMonitorRunnable.run` and `ServerMonitor.cancelCurrentCheck` to call `ServerMonitorRunnable.connection.close` concurrently.

But, there are other reasons in this class why `ServerMonitorRunnable.connection.close` may be called concurrently with itself: `ServerMonitor.close` may run concurrently with `ServerMonitorRunnable.run`, and it also may result in `ServerMonitorRunnable.connection.close` being called concurrently (the same reasoning applies to `RoundTripTimeRunnable.connection`). However, nothing was done in `DefaultServerMonitor` to prevent this, which may suggest that there was not intent to prevent `ServerMonitorRunnable.connection.close` from being called concurrently.

Overall, it is extremely difficult at times to reason about concurrency in the driver, because in many cases the thread-safety and related requirements/guarantees are not documented, especially for internal code. One of the outstanding culprits is `close` method on any interface/class. It is virtually never possible to say correctly or with confidence whether concurrent calls are allowed or not. The code we are discussing right now is a good example: one would expect that `InternalConnection` must never be used concurrently, yet, as I pointed out above, it seems like it may be, regardless of whether `synchronized` blocks are there or not.

TL;RD Thank you for pointing out the problem, the change may indeed introduce issues on top of those that `DefaultServerMonitor` seems to have anyway. I will replace `synchronized` critical sections with critical sections implemented using `DefaultServerMonitor.lock`.